File: 127-word-ladder/127-word-ladder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function(beginWord, endWord, wordList) {
  const alphabet = 'abcdefghijklmnopqrstuvwxyz';
  
  const words = new Set(wordList);
  let wordQueue = [beginWord];
  let depth = 1;
  
  while (wordQueue.length > 0) {
    const nextQueue = [];
    
    for (const word of wordQueue) {
      if (word === endWord) {
        return depth;
      }
      for (let i = 0; i < word.length; i++) {
        for (const c of alphabet) {
          const guess = buildGuess(word, i, c);
          
          if (words.has(guess)) {
            nextQueue.push(guess);
            words.delete(guess);
          }
        }
      }
    }
    
    depth += 1;
    wordQueue = nextQueue;
  }
  
  return 0;
};

const buildGuess = (word, i, c) => word.slice(0, i) + c + word.slice(i + 1);

File: 127-word-ladder/127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        words = set(wordList)
        q = deque([(beginWord, 1)])

        while q:
            word, depth = q.popleft()
            
            if word == endWord:
                return depth
            
            for guess in (
                ''.join((word[:i], c, word[i + 1:])) 
                for i in range(len(word)) 
                for c in ascii_lowercase
            ):
                if guess in words:
                    q.append((guess, depth + 1))
                    words.remove(guess)
        
        return 0